test/smc.jl

```diff
@@ -1,179 +1,153 @@
-using Turing, Random, Test
-using Turing.Core: ResampleWithESSThreshold
-using Turing.Inference: getspace, resample_systematic, resample_multinomial
-
-using Random
-
-dir = splitdir(splitdir(pathof(Turing))[1])[1]
-include(dir*"/test/test_utils/AllUtils.jl")
-
-@testset "SMC" begin
-    @turing_testset "constructor" begin
-        s = SMC()
-        @test s.resampler == ResampleWithESSThreshold()
-        @test getspace(s) === ()
+@testset "smc.jl" begin
+    @testset "SMC constructor" begin
+        sampler = AdvancedPS.SMC(10)
+        @test sampler.nparticles == 10
+        @test sampler.resampler == AdvancedPS.ResampleWithESSThreshold()
+
+        sampler = AdvancedPS.SMC(15, 0.6)
+        @test sampler.nparticles == 15
+        @test sampler.resampler ===
+            AdvancedPS.ResampleWithESSThreshold(AdvancedPS.resample_systematic, 0.6)
+
+        sampler = AdvancedPS.SMC(20, AdvancedPS.resample_multinomial, 0.6)
+        @test sampler.nparticles == 20
+        @test sampler.resampler ===
+            AdvancedPS.ResampleWithESSThreshold(AdvancedPS.resample_multinomial, 0.6)
+
+        sampler = AdvancedPS.SMC(25, AdvancedPS.resample_systematic)
+        @test sampler.nparticles == 25
+        @test sampler.resampler === AdvancedPS.resample_systematic
+    end
 
-        s = SMC(:x)
-        @test s.resampler == ResampleWithESSThreshold()
-        @test getspace(s) === (:x,)
+    # Smoke tests
+    @testset "models" begin
+        mutable struct NormalModel <: AbstractMCMC.AbstractModel
+            a::Float64
+            b::Float64
 
-        s = SMC((:x,))
-        @test s.resampler == ResampleWithESSThreshold()
-        @test getspace(s) === (:x,)
+            NormalModel() = new()
+        end
 
-        s = SMC(:x, :y)
-        @test s.resampler == ResampleWithESSThreshold()
-        @test getspace(s) === (:x, :y)
+        function (m::NormalModel)()
+            # First latent variable.
+            m.a = a = rand(Normal(4, 5))
```

Comment:
Maybe consider passing `rng` into the model?

Reply:
I tried this in my local branch when I realized that currently RNGs are not handled corrected by `Trace`. However, I reverted this change when I realized that this would require significant changes to `Trace`.

Reply:
See also https://github.com/TuringLang/AdvancedPS.jl/issues/21#issuecomment-813342838.

Reply:
Happy to address this in a separate PR (#23)